README.md
# smart-bartender
This is a Raspberry Pi smart bartender

## Installation

* make sure you have Python 3 installed
* create a virtual environment on your raspberry pi
* copy all data in the virtual environment folder
```
pip3 install -r /path/to/requirements.txt
```
* run app.py

done!


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

## [0.2.3] - 2018-09-15
- Adding a screen keyboard to the local device (so only to the Rpi-Screen.)
- !!! This includes a heavy bootstrap package and the keyboard extension is under GPLv3. -> needs to change for commercial use!!!

## [0.2.2] - 2018-09-11
- Using json instead of pickle for saving config. -> Class updates possible.

## [0.2.1] - 2018-09-10
- The Swap Bottle page now includes information regarding the filling level.

## [0.2.0] - 2018-09-10
- Initial release

[0.2.0]: https://github.com/fjfricke/smart-bartender/commits/master
[0.2.1]: https://github.com/fjfricke/smart-bartender/commit/6b374fa682c85079bcfddbedbe263aa33c044b30
[0.2.2]: https://github.com/fjfricke/smart-bartender/commit/aff4d3547d87cc428a5e4f1fd6199d13c5bd24fa
[0.2.3]: https://github.com/fjfricke/smart-bartender/commit/15e2392702e04bd765004e6fa0ae8012c352be0e


bar.py
import time
import sys
#import RPi.GPIO as GPIO
from RPi import GPIO
import json
from threading import Thread
import traceback
import pickle


class Bar:

    class Drink:

        class Version:

            def __init__(self, shortname: str, ingredients: [object], proportions: [float], glassize: int) -> object:
                self.shortname = shortname
                self.glassize = glassize

                assert (len(ingredients) == len(proportions)), "No. of drinks and no. of proportions must be equal."

                self.ingredients = ingredients  # list of Ingredient-Instances
                self.proportions = proportions  # list of distribution

                # for ingredient in ingredients:
                #    ingredient.assignTo(self)

            @classmethod
            def by_mililiters(cls, shortname, ingredients, volumina):
                # self.shortname = shortname
                assert (len(ingredients) == len(volumina)), "No. of drinks and no. of proportions must be equal."
                glassize = 0
                for volume in volumina:
                    glassize += volume
                proportions = []
                for volume in volumina:
                    proportions.append(volume / glassize * 100)
                return cls(shortname, ingredients, proportions, glassize)

            @property
            def available(self):
                for ingredient, proportion in zip(self.ingredients, self.proportions):
                    need = proportion / 100 * self.glassize
                    if not ingredient.available(need):
                        return False
                return True

            #    def make(self): (deprecated)
            #        # To Do
            #        threads = []
            #        needs = []
            #        for ingredient, proportion in zip(self.ingredients, self.proportions):
            #            need = proportion / 100 * self.glassize
            #            t = Thread(target=ingredient.pump.pour, args=(need,))
            #            t.start()
            #            threads.append(t)
            #            needs.append(need)
            #        while threads:
            #            for thread, ingredient, need in zip(threads, self.ingredients, needs):
            #                if thread.is_alive():
            #                    continue
            #                threads.remove(thread)
            #                ingredient.use(need)

            def make(self):
                # To Do
                threads = []
                for ingredient, proportion in zip(self.ingredients, self.proportions):
                    need = proportion / 100 * self.glassize
                    t = Thread(target=ingredient.pump.pour, args=(need,))
                    t.start()
                    threads.append(t)
                    ingredient.use(need)

            @property
            def duration(self):
                duration = 0
                for ingredient, proportion in zip(self.ingredients, self.proportions):
                    need = proportion / 100 * self.glassize
                    duration = max(duration, ingredient.pump.duration(volume=need))
                return duration

            def change_ingredients(self, ingredients: [object], proportions: [float]):

                assert (len(ingredients) == len(proportions)), "No. of drinks and no. of proportions must be equal."

                self.ingredients = ingredients  # list of Ingredient-Instances
                self.proportions = proportions  # list of distribution

        def __init__(self, name: str, type: str, picture: str, description: str) -> object:
            self.name = name
            self.type = type
            self.picture = picture  # name of picture file
            self.description = description
            self.versions = []

        def create_version(self, shortname: str, ingredients: [object], proportions: [float], glassize: int):
            version = self.Version(shortname, ingredients, proportions, glassize)
            self.versions.append(version)

        def delete_version(self, versionid):
            del self.versions[versionid]

        @property
        def available(self):
            for version in self.versions:
                if version.available:
                    return True
                return False

    class Ingredient:

        # available = false # => muss in Methode

        def __init__(self, name: str, percentage: float, bottlesize: float, remaining: float = None) -> object:
            self.name = name
            self.percentage = percentage  # in 0 - 100
            self.bottlesize = bottlesize  # in ml
            if remaining is None:
                self.remaining = bottlesize
            else:
                self.remaining = remaining  # in ml
            self.assignedTo = []

        def assignPump(self, pump: object):
            self.pump = pump

        def unplug(self):
            self.pump = None

        def available(self, need=0):
            if self.pump is None:
                return False
            if self.remaining - need < 0:
                return False
            return True

        def use(self, amount):
            self.remaining = self.remaining - amount

        def swapBottle(self):
            self.remaining = self.bottlesize

        def assignTo(self, version: object):
            self.assignedTo.append(version)

        def deleteFrom(self, version: object):
            self.assignedTo.remove(version)

        def change_Ingredient(self, name: str, percentage: float, bottlesize: float, pump: object = None, remaining: float = None):
            self.name = name
            self.percentage = percentage
            self.bottlesize = bottlesize
            if remaining is None:
                self.remaining = bottlesize
            else:
                self.remaining = remaining  # in ml
            if pump is None:
                self.unplug()
            else:
                self.assignPump(pump)

    # follow_up: method for new bottle, state of bottle, auto availability



    class Pump:

        def __init__(self, pin: int, flowrate: int = 500) -> object:
            #self.number = number
            self.pin = pin
            self.flowrate = flowrate  # standard: 500 ml/min

        def pour(self, volume):
            duration = volume / self.flowrate * 60
            GPIO.output(self.pin, GPIO.LOW)
            print('Start pouring from pin {} ...'.format(self.pin))
            time.sleep(duration)
            
            GPIO.output(self.pin, GPIO.HIGH)
            print('Stop pouring from pin {}, time elapsed: {}s.'.format(self.pin, duration))

        def duration(self, volume):
            return volume / self.flowrate * 60

        def clean(self):
            GPIO.output(self.pin, GPIO.LOW)
            print('Start cleaning pin {} ...'.format(self.pin))
            time.sleep(10)
            GPIO.output(self.pin, GPIO.HIGH)
            print('Stop cleaning pin {} ...'.format(self.pin))

    def __init__(self, folder:str = None):
        # check if saved drinks, ingredients, pumps, versions
        # otherwise create empty objects
        self.drinks = []
        self.ingredients = []
        self.pumps = []
        if folder:
            print("Reloading config from json..")
            self.reload_config(folder)

    def create_drink(self, name: str, type: str, picture: str, description: str):
        drink = self.Drink(name, type, picture, description)
        self.drinks.append(drink)

    def delete_drink(self, drink):
        self.drinks.remove(drink)

    def create_ingredient(self, name: str, percentage: float, bottlesize: float, remaining: float = None, pump: int = None):
        ingredient = self.Ingredient(name, percentage, bottlesize, remaining)
        if pump != None:
            ingredient.assignPump(self.pumps[pump])
        self.ingredients.append(ingredient)

    def pos_delete_ingredient(self, ingredient):
        for drink in self.drinks:
            for version in drink.versions:
                for assignedIng in version.ingredients:
                    if ingredient == assignedIng:
                        return False
        return True

    def delete_ingredient(self, ingredient):
        for drink in self.drinks:
            for version in drink.versions:
                for assignedIng in version.ingredients:
                    if ingredient == assignedIng:
                        return "assigned"
        self.ingredients.remove(ingredient)
        return 0

    def create_pump(self, pin: int, flowrate: int = 500):
        pump = self.Pump(pin, flowrate)
        self.pumps.append(pump)

    def pos_delete_pump(self, pump):
        for ingredient in self.ingredients:
            if ingredient.pump == pump:
                return False
        return True

    def delete_pump(self, pump: object):
        if not self.pos_delete_ingredient(pump):
            return "assigned"
        self.pumps.remove(pump)
        return 0

    def save_config(self, folder):

        json_pumps = []
        json_ingredients = []
        json_drinks=[]

        for pump in self.pumps:
            json_pumps.append({
                "pin": pump.pin,
                "flowrate": pump.flowrate
            })
        for ing in self.ingredients:
            dic = {
                "name": ing.name,
                "percentage": ing.percentage,
                "bottlesize": ing.bottlesize,
                "remaining": ing.remaining,
                "pump": self.pumps.index(ing.pump)
            }
            json_ingredients.append(dic)
        for drink in self.drinks:
            dic = {
                "name": drink.name,
                "type": drink.type,
                "picture": drink.picture,
                "description": drink.description
            }
            versions = []
            for version in drink.versions:
                vdic = {
                    "shortname": version.shortname,
                    "glassize": version.glassize,
                    "proportions": version.proportions
                }
                ingredients = []
                for ing in version.ingredients:
                    ingredients.append(self.ingredients.index(ing))
                vdic["ingredients"] = ingredients
                versions.append(vdic)
            dic["versions"] = versions
            json_drinks.append(dic)
        with open(folder + "/pumps.json", "w") as write_file:
            json.dump(json_pumps, write_file, indent=4, sort_keys=True)
        with open(folder + "/ingredients.json", "w") as write_file:
            json.dump(json_ingredients, write_file, indent=4, sort_keys=True)
        with open(folder + "/drinks.json", "w") as write_file:
            json.dump(json_drinks, write_file, indent=4, sort_keys=True)


    def reload_config(self, folder):
        try:
            pumps = []
            ingredients = []
            drinks = []

            with open(folder + "/pumps.json", "r") as read_file:
                pumps = json.load(read_file)
            print(pumps)
            with open(folder + "/ingredients.json", "r") as read_file:
                ingredients = json.load(read_file)
            print(ingredients)
            with open(folder + "/drinks.json", "r") as read_file:
                drinks = json.load(read_file)
            print(drinks)
            print("Reloading pumps..")
            for pump in pumps:
                self.create_pump(pin=pump["pin"], flowrate=pump["flowrate"])
            print("Reloading bottles..")
            for ing in ingredients:
                self.create_ingredient(name=ing["name"], percentage=ing["percentage"], bottlesize=ing["bottlesize"],
                                       remaining=ing["remaining"], pump=ing["pump"])
            print("Reloading drinks..")
            for iddrink, drink in enumerate(drinks):
                self.create_drink(name=drink["name"], type=drink["type"],
                                  picture=drink["picture"], description=drink["description"])
                versions = drink["versions"]
                for version in versions:
                    version_ings = []
                    for ingid in version["ingredients"]:
                        version_ings.append(self.ingredients[ingid])
                    self.drinks[iddrink].create_version(shortname=version["shortname"], ingredients=version_ings,
                                                        proportions=version["proportions"], glassize=version["glassize"])
            print("Reloading done!")
            return 0
        except:
            print("Could not reload from json. Something went wrong..")
            return 1

static/js/jquery-parent-to-animate.js
// Usage:
// $('.box').parentToAnimate($('.new-parent'), 200);
// $('.box').parentToAnimate($('.new-parent'), 'slow');
// $('.box').parentToAnimate('.new-parent', 'slow');

jQuery.fn.extend({
    // Modified and Updated by MLM
    // Origin: Davy8 (http://stackoverflow.com/a/5212193/796832)
    parentToAnimate: function(newParent, duration) {
        duration = duration || 'slow';
        
        var $element = $(this);
        
        newParent = $(newParent); // Allow passing in either a JQuery object or selector
        var oldOffset = $element.offset();
        $(this).appendTo(newParent);
        var newOffset = $element.offset();
        
        var temp = $element.clone().appendTo('body');
        
        temp.css({
            'position': 'absolute',
            'left': oldOffset.left,
            'top': oldOffset.top,
            'zIndex': 1000
        });
        
        $element.hide();
            
        temp.animate({
            'top': newOffset.top,
            'left': newOffset.left
        }, duration, function() {
            $element.show();
            temp.remove();
        });
    }
});

app.py
from flask import Flask
from flask import render_template, redirect, url_for
from flask import request
from flask import jsonify

from werkzeug.utils import secure_filename
import os

from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from threading import Thread

from bar import Bar

UPLOAD_FOLDER = os.path.dirname(__file__) + '/static/pictures/temp/'
PICTURE_FOLDER = os.path.dirname(__file__) + '/static/pictures/'
rel_folder = 'pictures/temp/'
ALLOWED_EXTENSIONS = set(['png', 'jpg', 'jpeg', 'gif'])

print(PICTURE_FOLDER)
print(__name__)
app = Flask(__name__)
app.config['UPLOAD_FOLDER'] = UPLOAD_FOLDER

'''BARTENDER2 SECTION'''
'''
bar = Bar()
bar.create_pump(11)
bar.create_pump(14, flowrate=1000)
bar.create_pump(17)
bar.create_pump(23)
bar.create_pump(25)

bar.create_ingredient(name="Wodka", percentage=45, bottlesize=700)
bar.ingredients[0].assignPump(bar.pumps[0])

bar.create_ingredient(name="Ginger Beer", percentage=0, bottlesize=500)
bar.ingredients[1].assignPump(bar.pumps[1])

bar.create_ingredient(name="Gin", percentage=45, bottlesize=700)
bar.ingredients[2].assignPump(bar.pumps[3])

bar.create_ingredient(name='Red Bull', percentage=0, bottlesize=200)
bar.ingredients[3].assignPump(bar.pumps[2])

bar.create_ingredient(name='Tonic Water', percentage=0, bottlesize=500)
bar.ingredients[4].assignPump(bar.pumps[4])

bar.create_drink(name="Moscow Mule", type='Longdrink', picture='moscowMule.png', description="This is a drink from Moscow.")
bar.drinks[0].create_version(shortname='single', ingredients=[bar.ingredients[0], bar.ingredients[1]], proportions=[40,60], glassize=200)
bar.drinks[0].create_version(shortname='double', ingredients=[bar.ingredients[0], bar.ingredients[1]], proportions=[50,50], glassize=200)
bar.drinks[0].create_version(shortname='big', ingredients=[bar.ingredients[0], bar.ingredients[1]], proportions=[60,40], glassize=8000)

bar.create_drink(name='Wodka Bull', type='Longdrink', picture='wodkaEnergy.png', description='This is a drink from Austria.')
bar.drinks[1].create_version(shortname='single', ingredients=[bar.ingredients[0], bar.ingredients[3]], proportions=[30,70], glassize=200)

bar.create_drink(name='Gin Tonic', type='Longdrink', picture='ginTonic.png', description='This is a fancy hyped drink.')
bar.drinks[2].create_version(shortname='single', ingredients=[bar.ingredients[2], bar.ingredients[4]], proportions=[30,70], glassize=200)



bar.create_pump(27)
bar.create_pump(32)

bar.create_ingredient(name="Liqueur 43", percentage=31, bottlesize=700)
bar.ingredients[5].assignPump(bar.pumps[5])

bar.create_ingredient(name="Milch", percentage=0, bottlesize=1000)
bar.ingredients[6].assignPump(bar.pumps[6])

bar.create_drink(name="Milch 43", type='Longdrink', picture='milch43.png', description="This is Vreni's and Felix' drink ❤️.")

bar.drinks[3].create_version(shortname='single', ingredients=[bar.ingredients[5], bar.ingredients[6]], proportions=[30,70], glassize=200)

bar.save_config('config')
'''

'''BARTENDER2 SECTION END'''

# bar = reload_config('config')
bar = Bar(folder='config')
myIP = '127.0.0.1'

def isRpi():
    if request.environ.get('HTTP_X_REAL_IP', request.remote_addr) == myIP:
        return True
    return False

@app.route("/")
def startpage():
    bar.save_config('config')
    if request.environ.get('HTTP_X_REAL_IP', request.remote_addr) == myIP:
        print('This is the Raspberry device!')
    return render_template('index.html', drinks=bar.drinks)

@app.route("/detail", methods=['GET'])
def detailview():
    iddrink = int(request.args.get('iddrink'))
    print(iddrink)
    drink = bar.drinks[iddrink]

    return render_template('detail.html', drink=drink, iddrink=iddrink)

@app.route("/make", methods=['POST'])
def makeview():
    iddrink = int(request.form.get('iddrink'))
    idversion = int(request.form.get('idversion'))
    duration = bar.drinks[iddrink].versions[idversion].duration
    bar.drinks[iddrink].versions[idversion].make()
    bar.save_config('config')
    return jsonify({'text': duration})

@app.route("/ingredients", methods=['GET','POST'])
def ingredientsview():
    if request.form.get('toast'):
        toast = request.form.get('toast')
    else:
        toast = ''
    print(isRpi())
    return render_template('/ingredients.html', ingredients=bar.ingredients, pumps=bar.pumps, bar=bar, toast=toast, isRpi=isRpi())

@app.route("/ingredients/changeBottleForm", methods=['POST'])
def changeBottleFormSnippet():
    idingredient = int(request.form.get('idingredient'))
    return render_template('snippets/changeBottleForm.html', idingredient=idingredient, ingredient=bar.ingredients[idingredient], pumps=bar.pumps, bar=bar, isRpi=isRpi())

@app.route("/submit/newIngredient", methods=['POST'])
def submitNewIngredient():
    name = request.form.get('name')
    pump = request.form.get('pump')
    percentage = int(request.form.get('percentage'))
    bottlesize = int(float(request.form.get('bottlesize')) * 1000)

    if pump == "":
        pumpid = None
    else:
        pumpid = int(pump)
    bar.create_ingredient(name=name, percentage=percentage, bottlesize=bottlesize, pump=pumpid)
    bar.save_config('config')
    return jsonify({'error': 0})

@app.route("/submit/changeIngredient", methods=['POST'])
def submitChangeIngredient():
    idingredient = int(request.form.get('idingredient'))
    name = request.form.get('name')
    pumpid = request.form.get('pump')
    percentage = int(request.form.get('percentage'))
    bottlesize = int(float(request.form.get('bottlesize')) * 1000)

    print(str(idingredient)+ " " + name + " " + pumpid + " " + str(percentage) + " " + str(bottlesize))

    if pumpid == "":
        pump = None
    else:
        pump = bar.pumps[int(pumpid)]

    bar.ingredients[idingredient].change_Ingredient(name=name, pump=pump, percentage=percentage, bottlesize=bottlesize)
    bar.save_config('config')
    return jsonify({'error': 0})

@app.route("/submit/deleteIngredient", methods=['POST'])
def submitDeleteIngredient():

    idingredient = int(request.form.get('idingredient'))
    bar.delete_ingredient(ingredient=bar.ingredients[idingredient])
    bar.save_config('config')
    return jsonify({'error': 0})

@app.route("/pumps", methods=['GET','POST'])
def pumpsview():
    if request.form.get('toast'):
        toast = request.form.get('toast')
    else:
        toast = ''
    return render_template('/pumps.html', pumps=bar.pumps, bar=bar, toast=toast, isRpi=isRpi())

@app.route("/pumps/changePumpForm", methods=['POST'])
def changePumpFormSnippet():
    idpump = int(request.form.get('idpump'))
    return render_template('snippets/changePumpForm.html', idpump=idpump, pump=bar.pumps[idpump], bar=bar, isRpi=isRpi())

@app.route("/submit/newPump", methods=['POST'])
def submitNewPump():
    pin = int(request.form.get('pin'))
    flowrate = request.form.get('flowrate')

    if flowrate == "":
        bar.create_pump(pin=pin)
    else:
        bar.create_pump(pin=pin, flowrate=int(flowrate))
    bar.save_config('config')
    return jsonify({'error': 0})

@app.route("/submit/changePump", methods=['POST'])
def submitChangePump():
    try:
        pin = int(request.form.get('pin'))
        flowrate = request.form.get('flowrate')
        idpump = int(request.form.get('idpump'))

        if not flowrate == "":
            bar.pumps[idpump].flowrate = int(flowrate)
        bar.pumps[idpump].pin = pin
        bar.save_config('config')
        return jsonify({'error': 0})
    except:
        return jsonify({'error': 1})

@app.route("/submit/deletePump", methods=['POST'])
def submitDeletePump():
    idpump = int(request.form.get('idpump'))
    bar.delete_pump(bar.pumps[idpump])
    bar.save_config('config')
    return jsonify({'error': 0})



@app.route("/drinks", methods=['GET', 'POST'])
def drinks(toast=''):
    if request.form.get('toast'):
        toast = request.form.get('toast')
    elif toast != '':
        toast = toast
    else:
        toast = ''
    filelist = os.listdir(PICTURE_FOLDER)
    for fichier in filelist[:]:  # filelist[:] makes a copy of filelist.
        if not (fichier.endswith(".png")):
            filelist.remove(fichier)
    print(filelist)
    return render_template('/drinks/drinks.html', toast=toast, pictures=filelist, drinks=bar.drinks, isRpi=isRpi())

@app.route("/drinks/chooseImage")
def cropperChooseImage():
    return render_template('drinks/chooseImage.html', isRpi=isRpi())

def allowed_file(filename):
    return '.' in filename and \
           filename.rsplit('.', 1)[1].lower() in ALLOWED_EXTENSIONS

@app.route("/drinks/cropImage", methods=['POST'])
def cropperCropImage():
    if request.method == 'POST':
        if request.form.get('path'):
            print(request.form.get('path'))
            return render_template('drinks/cropImage.html', path=request.form.get('path'))
        # check if the post request has the file part
        if 'file' not in request.files:
            print('No file part')
            return render_template('drinks/drinks.html')
        file = request.files['file']
        # if user does not select file, browser also
        # submit a empty part without filename
        if file.filename == '':
            print('No selected file')
            return render_template('drinks/drinks.html')
        if file and allowed_file(file.filename):
            print("works")
            filename = secure_filename(file.filename)
            file.save(os.path.join(app.config['UPLOAD_FOLDER'], filename))
            return render_template('drinks/drinks.html', path=rel_folder+filename)

    return render_template('drinks/drinks.html')
@app.route("/submit/newDrink", methods=['POST'])
def submitNewDrink():
    try:
        name = request.form.get('name')
        type = request.form.get('type')
        desc = request.form.get('desc')
        pic = request.form.get('picture')

        if name == '' or type == '' or pic == '':
            raise Exception()

        print(name + " " + type + " " + desc + " " + pic)

        bar.create_drink(name=name, type=type, picture=pic, description=desc)
        bar.save_config('config')
        return drinks('Successfully added new drink!')
    except:
        return drinks('Something went wrong..')

@app.route("/submit/deleteDrink", methods=['POST'])
def submitDeleteDrink():
    try:
        iddrink = request.form.get('iddrink')

        if iddrink == "":
            raise Exception()
        iddrink = int(iddrink)

        bar.delete_drink(drink=bar.drinks[iddrink])
        bar.save_config('config')
        return drinks('Successfully delete drink!')
    except:
        return drinks('Something went wrong..')

@app.route("/drinks/changeDrinkForm", methods=['POST'])
def changeDrinkFormSnippet():
    iddrink = int(request.form.get('iddrink'))

    filelist = os.listdir(PICTURE_FOLDER)
    for fichier in filelist[:]:  # filelist[:] makes a copy of filelist.
        if not (fichier.endswith(".png")):
            filelist.remove(fichier)

    return render_template('drinks/changeDrinkForm.html', iddrink=iddrink, drink=bar.drinks[iddrink], pictures=filelist, bar=bar, isRpi=isRpi())

@app.route("/submit/changeDrink", methods=['POST'])
def submitChangeDrink():
    try:
        iddrink = int(request.form.get('iddrink'))
        name = request.form.get('name')
        type = request.form.get('type')
        desc = request.form.get('desc')
        pic = request.form.get('picture')

        if name == '' or type == '' or pic == '':
            raise Exception()

        print(name + " " + type + " " + desc + " " + pic)

        bar.drinks[iddrink].name = name
        bar.drinks[iddrink].type = type
        bar.drinks[iddrink].picture = pic
        bar.drinks[iddrink].description = desc
        bar.save_config('config')
        return drinks('Successfully changed drink!')
    except:
        return drinks('Something went wrong..')

@app.route("/submit/newVersion", methods=['POST'])
def submitNewVersion():
    try:
        iddrink = request.form.get('iddrink')
        shortname = request.form.get('shortname')
        size = request.form.get('size')
        ings = request.form.getlist('ing[]')
        props = request.form.getlist('prop[]')

        if iddrink == "" or shortname == "" or size == "" or len(ings) == 0 or len(props) == 0:
            return drinks('You have to fillout all fields.')

        iddrink = int(iddrink)
        size = int(size)

        ingredients = []
        proportions = []
        for ing, prop in zip(ings, props):
            ingredients.append(bar.ingredients[int(ing)])
            proportions.append(int(prop))

        print(proportions)
        print(ingredients)

        bar.drinks[iddrink].create_version(shortname=shortname, ingredients=ingredients, proportions=proportions, glassize=size)
        bar.save_config('config')
        return drinks('Successfully added new drink!')
    except:
        return drinks('Something went wrong..')

@app.route("/submit/deleteVersion", methods=['POST'])
def submitDeleteVersion():
    try:
        iddrink = request.form.get('iddrink')
        idversion = request.form.get('idversion')

        if iddrink == "" or idversion == "":
            raise Exception()
        iddrink = int(iddrink)
        idversion = int(idversion)

        print(iddrink)
        print(idversion)

        bar.drinks[iddrink].delete_version(versionid=idversion)
        bar.save_config('config')
        return drinks('Successfully delete version!')
    except:
        return drinks('Something went wrong..')

@app.route("/swapping")
def swapView(toast=""):
    if request.method == 'GET':
        toast = request.args.get('toast')
    return render_template('/swapping.html', toast=toast, ingredients=bar.ingredients)

@app.route("/tools/swap", methods=['GET','POST'])
def swapAction():
    try:
        iding = "";
        if request.method == 'POST':
            iding = request.form.get('iding')
        else:
            iding = request.args.get('iding')

        if iding == "":
            raise Exception()

        iding = int(iding)

        bar.ingredients[iding].swapBottle()
        bar.save_config('config')
        return redirect(url_for('swapView', toast='Bottle refilled.'))
        return swapView('Bottle refilled.')
    except:
        return swapView('Something went wrong..')

@app.route("/cleaning")
def cleanView(toast=""):
    return render_template('/cleaning.html', toast=toast, pumps=bar.pumps)

@app.route("/tools/clean", methods=['POST'])
def cleanAction():
    try:
        iding = request.form.get('iding')

        if iding == "":
            raise Exception()
        iding = int(iding)

        bar.pumps[iding].clean()
        bar.save_config('config')
        return cleanView('Pump ' + str(iding+1) + ' cleaned.')
    except:
        return cleanView('Something went wrong..')

@app.route("/crop", methods=['POST'])
def crop():
    from PIL import Image

    path = request.form.get('path');
    coords = (float(request.form.get('x')), float(request.form.get('y')), float(request.form.get('x2')), float(request.form.get('y2')))
    width = float(request.form.get('w'))
    name = request.form.get('name')

    image_obj = Image.open('static/' + path)
    print(coords)
    cropped_image = image_obj.crop(coords)
    cropped_image.save('static/pictures/'+name)
    delpath = os.path.dirname(__file__) + "/static/" + path
    os.remove(delpath)

    return jsonify({'error': 0})

def startFirefox():
    driver = webdriver.Firefox()
    driver.get("http://127.0.0.1:5000")

def startChrome():
    chrome_options = Options()
    chrome_options.add_argument("--kiosk")
    driver = webdriver.Chrome(chrome_options=chrome_options)
    driver.get("http://127.0.0.1:5000")

if __name__ == "__main__":
    #t = Thread(target=startChrome)
    #t.start()
    #app.run(debug=True)
    myIP = '10.2.212.43'
    app.run(host=myIP, port=5000, debug=True)
    #app.run(host='10.2.211.39', port=5000, debug=True)
